src/tools/metrics/recorder.rs

```diff
@@ -0,0 +1,214 @@
+//! Metrics recording types and utilities.
+
+use std::collections::HashMap;
+
+use histogram::Histogram;
+use metrics::Key;
+use metrics_util::{
+    debugging::{DebugValue, DebuggingRecorder, Snapshotter},
+    CompositeKey, MetricKind,
+};
+
+// This is a false positive, `CompositeKey` does not depend on any interior mutable
+// types for its hashing implementation. Therefore, it is safe to use in our context.
+#[allow(clippy::mutable_key_type)]
+pub struct Snapshot(HashMap<CompositeKey, MetricVal>);
+
+impl Snapshot {
+    pub fn get_counter(&self, metric: &'static str) -> u64 {
```

Comment:
I'm not a rust expert but I'm wondering why not use here `String`? and when is appropriate to use `String` vs `&'static str`

Reply:
`String` has ownership over the contents of the string and is heap allocated, there's no need for that here. Since the metric name is a constant, hardcoding it into the executable and accessing it via a read only pointer to a string that'll be valid for the entire lifetime of the program seems appropriate?